### PR TITLE
refactor sig verify subcircuit

### PR DESCRIPTION
**Pull request by lispc**
TODOs:

1. constrain v == is_odd inside sig circuit && fix the sound issue of assert crt integer
2. connect tx circuit with sig table
3. the 1<<19 rows of sig circuit may not be optimal.. 1<<19 - 256? or something else?
4. heavy tests cost more time than before?
5. max_txs only used by tx circuit, and add 'max_sigs' or 'max_verif'? 

Some design discussions:
1. previously sig verification is embedded inside tx circuit, after this refactor, i also make a TxCircuitTester which include both  tx circuit and the new sig circuit, so mock-proving TxCircuitTester will also check signature verification. Is is good or we just stripping sig circuit out of tx circuit?
2. "enable-sig-verify" feature was used for historcial reasons when we did not have Ecdsa circuit. Now i think it is outdated and can be removed?
6. the instance column and "one" constant column in tx circuit,are they needed now?
7. is "onephase" feature in sig circuit needed?

### CODE REVIEW

File: zkevm-circuits/src/witness/block.rs [hunk 1]
```diff
@@ -100,6 +102,41 @@ impl<F: Field> Block<F> {
             }
         }
     }
+    pub(crate) fn get_sign_data(
+        &self,
+        padding: bool,
+    ) -> Result<Vec<SignData>, halo2_proofs::plonk::Error> {
```
Comment (roynalnaruto) [inline zkevm-circuits/src/witness/block.rs]: @lispc do we need this error? In the sig circuit this is called `unwrap` on. How about we filter only the items which are `Ok(Vec<SignData>)` instead, so this function becomes infallible?

Comment (roynalnaruto) [reply zkevm-circuits/src/witness/block.rs]: Also, I don't think we should fail here (if number of signatures are more than the `max_verif` config value. We can make this function infallible and later error on the synthesis step.

This function is only used in `new_from_block` which itself is infallible.

EDIT: Actually there is already a check for `signatures.len() > self.max_verif` in the `assign` method for `SigCircuit`. So this one is unnecessary I believe.

File: zkevm-circuits/src/sig_circuit.rs [hunk 2]
```diff
@@ -245,105 +188,105 @@ impl<F: Field> SignVerifyConfig<F> {
...
-            _marker: PhantomData::default(),
-            context_id: input.context_id,
-        }
+    fn synthesize_sub(
+        &self,
+        config: &Self::Config,
+        challenges: &Challenges<Value<F>>,
+        layouter: &mut impl Layouter<F>,
+    ) -> Result<(), Error> {
+        config.ecdsa_config.range.load_lookup_table(layouter)?;
```
Comment (zhenfeizhang) [inline zkevm-circuits/src/sig_circuit.rs:230]: This is already loaded during config. do we need to load it again?

File: zkevm-circuits/src/sig_circuit.rs [hunk 3]
```diff
@@ -497,15 +487,13 @@ impl<F: Field> SignVerifyChip<F> {
...
-        let pk_hash = (!padding)
-            .then(|| {
-                let mut keccak = Keccak::default();
-                keccak.update(&pk_be);
-                let hash: [_; 32] = keccak.digest().try_into().expect("vec to array of size 32");
-                hash
-            })
-            .unwrap_or_default()
-            .map(|byte| Value::known(F::from(byte as u64)));
+        let pk_hash = {
```
Comment (zhenfeizhang) [inline zkevm-circuits/src/sig_circuit.rs]: simplify with `let pk_hash = keccak256(pk_be).iter().map(...)`

File: zkevm-circuits/src/sig_circuit.rs [hunk 4]
```diff
@@ -533,26 +521,35 @@ impl<F: Field> SignVerifyChip<F> {
...
-            .iter()
-            .map(|&x| QuantumCell::Witness(Value::known(F::from_u128(x as u128))))
-            .collect_vec();
+
+        let assert_crt = |ctx: &mut Context<F>,
+                          bytes: [u8; 32],
+                          crt_integer: &CRTInteger<'v, F>,
+                          overriding: &Option<&QuantumCell<F>>|
+         -> Result<_, Error> {
+            //let bytes: [u8; 32] = v.to_bytes();
```
Comment (zhenfeizhang) [inline zkevm-circuits/src/sig_circuit.rs]: remove

File: zkevm-circuits/src/sig_circuit.rs [hunk 5]
```diff
@@ -533,26 +521,35 @@ impl<F: Field> SignVerifyChip<F> {
...
+            let byte_cells: Vec<QuantumCell<F>> = bytes
+                .iter()
+                .map(|&x| QuantumCell::Witness(Value::known(F::from_u128(x as u128))))
```
Comment (zhenfeizhang) [inline zkevm-circuits/src/sig_circuit.rs]: change to `F::from(x as u64)` 
this allows for a bit faster allocation with witness-gen optimization

File: zkevm-circuits/src/sig_circuit.rs [hunk 6]
```diff
@@ -812,6 +842,46 @@ impl<F: Field> SignVerifyChip<F> {
...
+                for (idx, assigned_sig_verif) in assigned_sig_verifs.iter().enumerate() {
+                    region.assign_fixed(
+                        || "assign sig_table selector",
+                        config.sig_table.q_enable,
+                        idx,
+                        || Value::known(F::one()),
+                    )?;
+
+                    // FIXME: constrain v
+                    region.assign_advice(
```
Comment (zhenfeizhang) [inline zkevm-circuits/src/sig_circuit.rs]: `v` need to be copy-constrained

File: zkevm-circuits/src/sig_circuit/test.rs [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        max_verif,
+        signatures,
+        _marker: PhantomData,
+    };
+
+    let prover = match MockProver::run(k, &circuit, vec![vec![]]) {
+        Ok(prover) => prover,
+        Err(e) => panic!("{:#?}", e),
+    };
+    assert_eq!(prover.verify(), Ok(()));
```
Comment (zhenfeizhang) [inline zkevm-circuits/src/sig_circuit/test.rs:100]: The circuit no longer check if the signature is valid. Maybe add a test here?

File: zkevm-circuits/src/tx_circuit.rs [hunk 8]
```diff
@@ -142,7 +140,14 @@ pub struct TxCircuitConfig<F: Field> {
 
     /// Address recovered by SignVerifyChip
     sv_address: Column<Advice>,
-    sign_verify: SignVerifyConfig<F>,
+
+    /// A fixed column to store F::one, used to constraint signature_is_valid == one.
+    // FIXME: it is pretty wasteful to allocate a new fixed column just
+    // to store a constant value F::one.
+    // Since we will use lookup to constrain is_valid later, is the "constant one" fixed column
+    // still needed?
```
Comment (zhenfeizhang) [inline zkevm-circuits/src/tx_circuit.rs]: sounds good to remove :-)

File: zkevm-circuits/src/tx_circuit.rs [hunk 9]
```diff
@@ -177,11 +184,17 @@ impl<F: Field> SubCircuitConfig<F> for TxCircuitConfig<F> {
             block_table,
             keccak_table,
             rlp_table,
+            sig_table,
             challenges: _,
         }: Self::ConfigArgs,
     ) -> Self {
         let q_enable = tx_table.q_enable;
 
+        // not needed now?
```
Comment (zhenfeizhang) [inline zkevm-circuits/src/tx_circuit.rs]: sounds good to remove :-)

File: zkevm-circuits/src/witness/tx.rs [hunk 10]
```diff
@@ -120,6 +120,9 @@ impl Transaction {
 
     /// Sign data
     pub fn sign_data(&self) -> Result<SignData, Error> {
+        if self.r.is_zero() && self.s.is_zero() && self.v == 0 {
```
Comment (zhenfeizhang) [inline zkevm-circuits/src/witness/tx.rs:123]: https://en.wikipedia.org/wiki/Elliptic_Curve_Digital_Signature_Algorithm#Signature_verification_algorithm
>  Verify that r and s are integers in [ 1 , n − 1 ] [1,n-1]. If not, the signature is invalid.

change to
`self.r.is_zero || self.s.is_zero()` ?

File: zkevm-circuits/src/sig_circuit.rs (hunk 2 above)
Comment (lispc) [reply zkevm-circuits/src/sig_circuit.rs:230]: <del>oh i should do this in assignment. not in config.</del>

no i think it is not loaded during config?

File: zkevm-circuits/src/tx_circuit.rs [hunk 11]
```diff
@@ -1111,9 +1138,6 @@ impl<F: Field> TxCircuitConfig<F> {
             },
         )?;
 
-        #[cfg(feature = "enable-sign-verify")]
-        self.sign_verify.load_range(layouter)?;
```
Comment (zhenfeizhang) [inline zkevm-circuits/src/tx_circuit.rs:1115]: re: https://github.com/scroll-tech/zkevm-circuits/pull/538#discussion_r1233565808

the range config was loaded here... it seems it got removed during refactoring.

File: zkevm-circuits/src/sig_circuit.rs [hunk 12]
```diff
@@ -245,105 +184,104 @@ impl<F: Field> SignVerifyConfig<F> {
...
             q_keccak,
-            fixed_column,
             rlc_column,
         }
     }
 }
 
-impl<F: Field> SignVerifyConfig<F> {
+impl<F: Field> SigCircuitConfig<F> {
     pub(crate) fn load_range(&self, layouter: &mut impl Layouter<F>) -> Result<(), Error> {
```
Comment (zhenfeizhang) [inline zkevm-circuits/src/sig_circuit.rs:195]: also re: https://github.com/scroll-tech/zkevm-circuits/pull/538#discussion_r1233565808

File: mock/src/transaction.rs [hunk 13]
```diff
@@ -171,7 +171,8 @@ impl Default for MockTransaction {
             block_hash: Hash::zero(),
             block_number: U64::zero(),
             transaction_index: U64::zero(),
-            from: AddrOrWallet::Addr(MOCK_ACCOUNTS[0]),
+            //from: AddrOrWallet::Addr(MOCK_ACCOUNTS[0]),
+            from: AddrOrWallet::random(&mut OsRng),
```
Comment (roynalnaruto) [inline mock/src/transaction.rs:175]: why this change?

Comment (lispc) [reply mock/src/transaction.rs:175]: if `from` is  AddrOrWallet::Addr, the signature inside witness::tx will be. {r:0, s:0, v:0}.   Switching to AddrOrWallet::Wallet will make the tx more "real".
